File: part03/ch12/Q11.py
from collections import deque

n = int(input())
data = [[0] * (n+1) for _ in range(n+1)]

k = int(input())
for _ in range(k):
    a, b = map(int, input().split())
    data[a][b] = 1

order = []
l = int(input())
for _ in range(l):
    t, d = input().split()
    order.append((int(t), d))

snack = deque()
snack.append((1, 1))
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

dist = 0
result = 0
idx = 0
while True:
    nx = snack[-1][0] + dx[dist]
    ny = snack[-1][1] + dy[dist]
    result += 1
    if nx < 1 or ny < 1 or nx > n or ny > n:
        break
    if (nx, ny) in snack:
        break
    if data[nx][ny] == 1:
        data[nx][ny] = 0
    else:
        snack.popleft()
    snack.append((nx, ny))
    if idx < l:
        if order[idx][0] == result:
            if order[idx][1] == 'L':
                dist -= 1
                if dist == -1:
                    dist = 3
            else:
                dist += 1
                if dist == 4:
                    dist = 0
            idx += 1

print(result)

File: part03/ch12/Q11.cpp
#include <bits/stdc++.h>

using namespace std;

int n, k, l;
vector<pair<int, char>> dp;


int main() {
    int data[101][101] = {0};

    cin >> n;
    cin >> k;
    for (int i = 0; i < k; i++) {
        int a, b;
        cin >> a >> b;
        data[a][b] = 1;
    }
    cin >> l;
    for (int i = 0; i < l; i++) {
        int t;
        char c;
        cin >> t >> c;
        dp.push_back({t, c});
    }
    deque<pair<int, int>> snack;
    snack.push_back({1, 1});

    int dx[4] = {0, 1, 0, -1};
    int dy[4] = {1, 0, -1, 0};

    int snackIdx = 0;
    int dist = 0;
    int result = 0;
    int idx = 0;
    while (true) {
        int nx = snack.back().first + dx[dist];
        int ny = snack.back().second + dy[dist];
        result++;
        if (nx < 1 || ny < 1 || nx > n || ny > n) {
            cout << result << endl;
            return 0;
        }
        for (int i = 0; i < snack.size(); i++) {
            if (nx == snack[i].first && ny == snack[i].second) {
                cout << result << endl;
                return 0;
            }
        }
        if (data[nx][ny] == 1) {
            data[nx][ny] = 0;
        } else {
            snack.pop_front();
        }
        snack.push_back({nx, ny});
        if (idx < l) {
            if (result == dp[idx].first) {
                if (dp[idx].second == 'L') {
                    dist -= 1;
                    if (dist == -1) {
                        dist = 3;
                    }
                } else {
                    dist += 1;
                    if (dist == 4) {
                        dist = 0;
                    }
                }
                idx++;
            }
        }
    }
}
